### PR TITLE
FIX : Recursively retrieves all parent bom in the tree that leads to …

### PR DESCRIPTION
**Pull request by atm-gauthier**
…the $bom_id bom to check if we're allowed to add bom as bomline

### CODE REVIEW

File: htdocs/bom/class/bom.class.php [hunk 1]
```diff
@@ -1147,6 +1147,30 @@ public function getNetNeedsTree(&$TNetNeeds = array(), $qty = 0, $level = 0)
...
+	 * @param 	int		$bom_id			ID of bom from which we want to get parent bom ids
+	 * @return 	void
+	 */
+	public function getParentBomTreeRecursive(&$TParentBom, $bom_id)
+	{
+
+		$sql = 'SELECT l.fk_bom, b.label
+				FROM '.MAIN_DB_PREFIX.'bom_bomline l
+				INNER JOIN '.MAIN_DB_PREFIX.$this->table_element.' b ON b.rowid = l.fk_bom
+				WHERE fk_bom_child = '.$bom_id;
```
Comment (frederic34) [inline htdocs/bom/class/bom.class.php:1164]: @atm-gauthier add int cast here

File: htdocs/bom/class/bom.class.php [hunk 2]
```diff
@@ -1147,6 +1147,30 @@ public function getNetNeedsTree(&$TNetNeeds = array(), $qty = 0, $level = 0)
...
+	{
+
+		$sql = 'SELECT l.fk_bom, b.label
+				FROM '.MAIN_DB_PREFIX.'bom_bomline l
+				INNER JOIN '.MAIN_DB_PREFIX.$this->table_element.' b ON b.rowid = l.fk_bom
+				WHERE fk_bom_child = '.((int) $bom_id);
+
+		$resql = $this->db->query($sql);
+		if (!empty($resql)) {
+			while ($res = $this->db->fetch_object($resql)) {
```
Comment (eldy) [inline htdocs/bom/class/bom.class.php:1168]: Recursive methods always need a protection against infinite loop.
Imagine data was loaded with a child thas has a child that is its parent, you fall into an infinite loop.
You must add a counter that is increased +1 at each call and we stop the recursion when counter > 1000 (we know the depth of BOM will never be 1000).

Comment (atm-gauthier) [reply htdocs/bom/class/bom.class.php:1176]: Hi @eldy, i just did it

File: htdocs/bom/class/bom.class.php [hunk 3]
```diff
@@ -1147,6 +1147,38 @@ public function getNetNeedsTree(&$TNetNeeds = array(), $qty = 0, $level = 0)
...
+	 * @param 	int		$bom_id			ID of bom from which we want to get parent bom ids
+	 * @param 	int		$level		Protection against infinite loop
+	 * @return 	void
+	 */
+	public function getParentBomTreeRecursive(&$TParentBom, $bom_id = '', $level = 1)
+	{
+
+		// Protection against infinite loop
+		if ($level > 1000) {
+			return 0;
```
Comment (eldy) [inline htdocs/bom/class/bom.class.php:1164]: Not compatible with doc saying   "return void"
Just do "return" or better, change return code in doc into   @return  int

**Comment by atm-gauthier** [comment]
@eldy stickler gives an error but no information about it

**Comment by eldy** [comment]
> @eldy stickler gives an error but no information about it

Don't worry about sticker. Only Travis is important.

**Comment by atm-gauthier** [comment]
@eldy OK, is everything OK now with this pr for you?

**Pull request merged**